### DISCUSSION TITLE
supabase realtime or pg_graphql realtime vs graphql subscriptions or hasura

### DISCUSSION THREAD

**Discussion by jtkeyva** [supabase/supabase#9892] [Questions]
huge supabase fan with some objective questions:

why is supabase pg_graphql or supabase realtime better than graphql subscriptions or hasura?

what's the difference? hasura supports realtime + caching amongst other friendly features.

why would anyone use anything different?

**Comment by olirice** [comment]
> what's the difference? hasura supports realtime + caching amongst other friendly features.

realtime and hasura graphql subscriptions fill a similar niche. The biggest differences are

- security implementation

realtime uses postgres' builtin role permissions and row level security policies to control which users can see different tables/columns/rows. Those permissions and policies are shared by the supabase REST API, GraphQL API, and the SQL interface. You define your security using postgres primitives, everything on the platform obeys it. Leaning in to postgres' builtin security mechanisms also makes it portable if you ever want to switch to a new vendor

hasura has a proprietary security implementation that is configured outside the database. I'm not very familiar with the details so here's a [reference to their guide](https://hasura.io/learn/graphql/hasura-advanced/security/)
 
- non-graphql vs graphql

once you go graphql its easy to forget that not everyone has made the switch :)

side note, there is a plan for GraphQL subscriptions in pg_graphql too but its still a little ways out

> why is supabase pg_graphql or supabase realtime better than graphql subscriptions or hasura? ... why would anyone use anything different?

Hasura has a suite of great products. A key benefit of supabase being "just postgres" + add-ons is that you can pick and choose which components to use and it typically plays well with other tools. If you're as happy with Hasura as it sounds like you are, running that stack against your supabase DB is a totally reasonable choice

**Comment by jtkeyva** [comment]
thanks for the thorough explanation! look forward to subscriptions progress
